L73-java-binderchannel/wireformat.md

```diff
@@ -0,0 +1,173 @@
+```
+
+#### Flag fields
+
+*   FLAG_PREFIX - Indicates the transaction contains the streams prefix.
+*   FLAG_MESSAGE_DATA - Indicates the transaction contains message data.
+*   FLAG_SUFFIX - Indicate the transaction contains streams suffix.
+*   FLAG_OUT_OF_BAND_CLOSE - Indicates this transaction contains out of band
+    close data. (This type is only ever sent in a transaction by itself).
+*   FLAG_MESSAGE_DATA_IS_PARCELABLE - Indicates the message is a parcelable
+    object.
+*   FLAG_STATUS_DESCRIPTION - Indicates the status description string is
+    included.
+*   status - If a status is included in the data, the status code will be
+    present in the top 16 bits of the flags value.
+
+#### Sequence number
+
+The sequence number is used to detect dropped, or out of sequence transactions.
+It’s always 0 on the first transaction of a stream, and increments with each
+transaction sent. The sequence number will wrap around to 0 if more than 2^31
+messages are sent. The sequence numbers of client-to-server and server-to-client
+transactions are independent of each other.
+
+## Transport Setup
+
+The setup of a new transport occurs in the following steps.
+
+1.  The ClientTransport first binds to the endpoint Android Service, to obtain
+    an endpoint binder.
+2.  The ClientTransport sends a _setup transport_ transaction to the endpoint
+    binder, passing the client binder in the parcel, then discarding the
+    reference to the endpoint binder.
+3.  The server receives the _setup transport_ transaction from the client and
+    creates a ServerTransport instance, passing it a reference to the client
+    binder.
+4.  The newly-created ServerTransport instance sends a _setup transport_
+    transaction back to the client, passing the server binder in the parcel.
+5.  The ClientTransport receives the server binder.
+6.  The transport pair is now ready, and rpc transactions can be sent.
+
+## Transport Lifecycle & Shutdown
+
+Each transport implementation only connects once, with any failure causing it to
+shutdown for good. Reconnection attempts and retries are left up to the
+ManagedChannel which creates transports when needed. A transport can be shut
+down either gracefully, or ungracefully. A graceful shutdown causes all new
+calls to be rejected, but allows ongoing calls to complete normally.
+
+By comparison, an ungraceful shutdown closes all ongoing calls, sends a
+shutdown-transport transaction to the remote binder, and shuts down all binders
+immediately. If the remote binder dies, or any internal failure occurs, an
+ungraceful shutdown will occur. If the Android Service hosting the server is
+destroyed, a graceful shutdown will occur, though if the host process also dies,
+that will lead to a binder death in the client, shutting the client-side down
+ungracefully. If the Android component hosting the Channel is destroyed, an
+ungraceful shutdown will occur, with the status code ABORTED. The remote
```

Comment:
We can't generate ABORTED from within the library (see very end of [statuscodes.md](https://github.com/grpc/grpc/blob/master/doc/statuscodes.md)). This would generally be exposed as UNAVAILABLE or CANCELLED (depending on implementation-specific details).

Reply:
Cool, I'll go with CANCELLED since that seems the most appropriate.